cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java

```diff
@@ -200,15 +197,54 @@ private synchronized void performGroupBasedGrant(EntityId entityId, Principal pr
   }
 
   /**
-   * Drops a role if it was created by cdap i.e. it starts with ENTITY_ROLE_PREFIX if there is no privileges
+   * Drops a role if it was created by cdap i.e. it starts with ENTITY_ROLE_PREFIX and there is no privileges
    * associated with it
    * @param entityRole the role which needs to be dropped if empty
-   * @throws RoleNotFoundException if the given role does not exists
    */
-  private void cleanUpRole(Role entityRole) throws RoleNotFoundException {
-    if (entityRole.getName().startsWith(ENTITY_ROLE_PREFIX) && listPrivileges(entityRole).isEmpty()) {
+  private void cleanUpEntityRole(Role entityRole) {
+    // this should not be called for any other role except entity roles i.e. the ones which start with
+    // ENTITY_ROLE_PREFIX
+    if (!entityRole.getName().startsWith(ENTITY_ROLE_PREFIX)) {
+      throw new IllegalArgumentException(String.format("The given role %s is not an entity role. " +
+                                                         "Please use drop role to remove this role.", entityRole));
+    }
+    // if there are privileges associated with this role then don't clean it up
+    if (!listPrivileges(entityRole).isEmpty()) {
+      return;
+    }
+    try {
       binding.dropRole(entityRole);
+      LOG.debug("Successfully dropped role {}", entityRole);
+    } catch (RoleNotFoundException e) {
+      // This is a dot role. It should be ok for deletion to fail. This happens because while creating a new entity,
+      // we first revoke any orphaned privileges on the entity. During that operation this role may not exist.
+      LOG.debug("Trying to delete role {}, but it was not found. Ignoring since its an entity role.", entityRole);
+    }
+  }
+
+  /**
+   * Gets all the entity roles associated with the given {@link EntityId}
+   * @param entityId the entity for which roles need to be obtained
+   * @return {@link Set} of {@link Role} for the given entity
+   */
+  private Set<Role> getEntityRoles (final EntityId entityId) {
+    final String curEntityRolePrefix = Joiner.on(ENTITY_ROLE_PREFIX).join("", entityId.toString());
+
+    Predicate<Role> filter = new Predicate<Role>() {
+      public boolean apply(Role role) {
+        return role.getName().startsWith(curEntityRolePrefix);
+      }
+    };
+
+    // get all roles and filter roles which belongs to this this entity
+    Set<Role> existingRoles = listAllRoles();
```

Comment:
Can replace this with 
`return Sets.newHashSet(Collections2.filter(listAllRoles(), filter));`